### PR TITLE
some bulk sequence loading tests that nail down current ACGTN behavior.

### PR DESCRIPTION
**Pull request by ctb**
These tests nail down behavior prior to #1590, which, when merged, will alter how we handle non-ACTG characters.  Note, no behavior is changed in this PR; it's just (lots of) new tests.

This explicitly puts in place tests for sequences that contain one of lowercase, Ns, and non-ACGTN characters, for:

* `consume_fasta` and all other bulk-sequence loading functions on Hashtables and derived classes;
* `trim_on_abundance`, `trim_below_abundance` and `find_spectral_error_positions`

This PR includes #1661.

Adds new test file `tests/test_sequence_validation.py` and data file `tests/test-data/valid-read-testing.fq`.

----

- [x] Is it mergeable?
- [x] `make test` Did it pass the tests?
- [x] `make clean diff-cover` If it introduces new functionality in
  `scripts/` is it tested?
- [x] `make format diff_pylint_report cppcheck doc pydocstyle` Is it well
  formatted?
- [x] Did it change the command-line interface? Only backwards-compatible
  additions are allowed without a major version increment. Changing file
  formats also requires a major version number increment.
- [x] For substantial changes or changes to the command-line interface, is it
  documented in `CHANGELOG.md`? See [keepachangelog](http://keepachangelog.com/)
  for more details.
- [x] Was a spellchecker run on the source code and documentation after
  changes were made?
- [x] Do the changes respect streaming IO? (Are they
  tested for streaming IO?)

### CODE REVIEW

**Comment by ctb** [comment]
Wow, [these tests](https://github.com/dib-lab/khmer/blob/c0c5d2cb77639a7dc163618ffdb910f36afe670d/tests/test_read_parsers.py#L487) are like a showcase of horrible inconsistency in loading sequences.

File: tests/test_read_parsers.py [hunk 1]
```diff
@@ -483,5 +483,204 @@ def test_clean_seq():
...
+    x = Counttable(15, int(1e6), 4)
+    for read in ReadParser(infile):
+        x.consume(read.sequence)          # consume raw sequence
+
+    # the relevant read will be entirely ignored
+    # (b/c ReadParser does not uppercase)
+    kmer = "caggcgcccaccacc".upper()
+    assert x.get(kmer) == 1
+
+    # the 2nd read with this k-mer in it has an N in it; 'consume' will ignore.
```
Comment (betatim) [inline tests/test_read_parsers.py]: I'd rephrase the comment. At the moment I am very puzzled after reading it (ah its being ignored so count should be one, but it is asserted to be two ...??)

Comment (betatim) [inline tests/test_read_parsers.py]: Maybe "`consume` will ignore the invalid base and continue consuming the read, so this kmer after the N should have abundance 2"

**Comment by betatim** [comment]
I'd move these tests to a new file maybe `test_sequence_loading.py`? Then we have one file that deals exclusively with verifying the assumptions we make about who cleans what, when and how.

Because it is easy to do I would parametrise the tests (that make sense) on the class so we test all combinations of Count/Node and table/graph. In which case we should get the parametrisation stuff from `test_tabletype.py` (or put these tests there, I'm -0)

File: tests/test_read_parsers.py [hunk 2]
```diff
@@ -483,5 +483,204 @@ def test_clean_seq():
...
+    x.add_tag(kmer)
+    x.set_partition_id(kmer, 2)
+    
+    kmer = 'CCGGCGTG'
+    x.add_tag(kmer)
+    x.set_partition_id(kmer, 3)
+    
+    x.output_partitions(infile, savepath)
+
+    read_names = [ read.name for read in ReadParser(savepath) ]
```
Comment (betatim) [inline tests/test_read_parsers.py]: The `pep8` :police_car: doesn't like the extra white space (hence the travis failure)

**Comment by betatim** [comment]
+1 on merging this before 2.1, then #1590 can update the behaviour and fix these tests (remove the "in the future ..." comments)

File: tests/test_read_parsers.py (hunk 1 above)
Comment (ctb) [inline tests/test_read_parsers.py]: fixed in fe48787

File: tests/test_read_parsers.py (hunk 2 above)
Comment (ctb) [inline tests/test_read_parsers.py]: fixed 756454e

**Comment by ctb** [comment]
Tests moved to `test_sequence_validation.py` and parameterized as per @betatim suggestion! Of course now a bunch of tests are failing... :)

**Comment by betatim** [comment]
Linux build fails because of a pep8 violation. The OSX build fails because some tests fail and then we exit ungracefully because there are [too many open files](https://travis-ci.org/dib-lab/khmer/jobs/205207362#L3470). Will take a look at the latter.

**Comment by betatim** [comment]
Ha, locally on OSX all tests pass as well and no "too many open files" warnings.

**Comment by betatim** [comment]
Switched to using a fixture for `ReadParser` which allows us to explicitly close it after using it. Not sure this fixes the problem, but it is my best guess as I can't reproduce it locally.

**Comment by betatim** [comment]
🎉

**Comment by ctb** [comment]
This is now ready for review & merge (although #1661 should be merged first :)

@luizirber @betatim @standage

**Comment by ctb** [comment]
(Upon merge, we should switch #1590 over to be against master.)

File: tests/test_sequence_validation.py [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+        posns = x.find_spectral_error_positions(s, 1)
+    # assert posns == []                    # in future, should return [11]
+
+
+def test_read_cleaning_output_partitions(graphtype):
+    infile = utils.get_test_data('valid-read-testing.fq')
+    savepath = utils.get_temp_filename('foo')
+
+    # read this in using "approved good" behavior w/cleaned_seq
+    x = _Nodegraph(8, PRIMES_1m)
```
Comment (betatim) [inline tests/test_sequence_validation.py]: -> `graphtype`?

File: tests/test_sequence_validation.py [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+    return request.param
+
+
+# all the graph types!
+@pytest.fixture(params=[_Countgraph, _Nodegraph])
+def graphtype(request):
+    return request.param
+
+
+@pytest.fixture
```
Comment (betatim) [inline tests/test_sequence_validation.py]: Now that I've looked this up: can we change this back to `yield` and change the decorator to `@pytest.yield_fixture`? Link to the pytest 2.9 docs: http://doc.pytest.org/en/2.9.2/fixture.html#fixture-finalization-executing-teardown-code

**Comment by ctb** [comment]
done

**Comment by ctb** [comment]
all comments addressed in e419a7b

**Comment by betatim** [comment]
:tada:
